Submission logic base.

Mercurrent opened the pull request:
I refrained from adding more tests until we agree on basic logic.

Let's see if comment mention notifies @jdkasten

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	SleepTime = time.Second
+)
+
+type guardedLogs struct {
```
There's nothing log-specific about this, so the naming is misleading -- it describes how you happen to want to use it, rather than what it does.  AFAICT it's really a `safeStringSet` or somesuch?

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+func (gList *guardedLogs) add(logURL string) {
+	gList.m.Lock()
+	gList.logURLs[logURL] = true
+	gList.m.Unlock()
```
nit: usually better to `defer` the unlock:
```golang
gList.m.Lock()
defer gList.m.Unlock()
...
```
(passim)

Mercurrent replied:
Changed.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	// URLProtocol keeps protocol substring for Log URLs.
+	URLProtocol = "https://"
+
+	// Sleeptim is duration between parallel batch call and subsequent requests to Logs within group.
```
Comment doesn't match constant name, but the name is a bit too generic anyhow -- maybe something like `PostBatchInterval` ?

Mercurrent replied:
Done.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+const (
+	// URLProtocol keeps protocol substring for Log URLs.
+	URLProtocol = "https://"
```
This is unused and exported -- does it belong here?

Mercurrent replied:
Removed.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	return exist
+}
+
+func (gList *guardedLogs) sumFor(URLs map[string]bool) int {
```
Is this method needed?  Could a caller loop over their own URL map and just call `gList.has(entry)` for each one?

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	return entries
+}
+
+type result struct {
```
maybe `submissionResult` ?

Mercurrent replied:
?

daviddrysdale replied:
`result` is very generic -- result of what?  A terse name is fine if the scope of use of the type is short, but as the scope gets bigger the name needs to be more descriptive.  This type is package-private but is passed between goroutines, so is somewhere in the middle.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+type result struct {
+	sct    *ct.SignedCertificateTimestamp
+	logURL string
```
Ordering the fields `logURL`, `sct`, `err` would feel more natural to me.

Mercurrent replied:
reordered.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+				if err != context.Canceled {
+					// TODO(Mercurrent) Submission failed. Log it
+				}
+				results <- result{err: err}
```
Any reason not to populate `logURL` in the error result too?  Seems like it might be helpful to have.

Mercurrent replied:
Nice catch.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToSingleCTWithResult(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
```
The name seems a bit superfluous -- the `Single` and `WithResult` parts are clear from the method signature.  Maybe `SubmitToLog()` ?

Mercurrent replied:
Done.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	results chan result, requestList *guardedLogs, sctList *guardedLogs, submitter Submitter) {
+	groupURLs := reflect.ValueOf(group.LogURLs).MapKeys()
```
Using `reflect` should be a last resort, I don't think you need it here.

Mercurrent replied:
Removed.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	results chan result, requestList *guardedLogs, sctList *guardedLogs, submitter Submitter) {
+	groupURLs := reflect.ValueOf(group.LogURLs).MapKeys()
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
```
If you're spinning up parallel goroutines for the N URLs, is the `rand.Perm` necessary?

Mercurrent replied:
Added comment.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	return b
+}
+
+// sleepTime calculates duration for consequent call. For first paralleStart calls duration is 0, while every next one gets additional dur interval.
```
s/paralleStart/parallelStart/

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+// sleepTime calculates duration for consequent call. For first paralleStart calls duration is 0, while every next one gets additional dur interval.
+func sleepTime(idx int, parallelStart int, dur time.Duration) time.Duration {
+	return (time.Duration(max(idx+1-parallelStart, 0)) * dur)
```
I think this would be easier to follow if the `max` logic were folded in here:
```golang
  if idx < parallelStart {
    return time.Duration(0)
  }
  return time.Duration(idx+1-parallelStart)*dur
```

Mercurrent replied:
Done.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+			if requestList.has(logURL) {
+				return
+			}
+			requestList.add(logURL)
```
This is racy -- another goroutine could `add` between the calls to `has` and `add`.  Better to add something test-and-set like on the `guardedLogs` type.

Mercurrent replied:
Changed logic.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	results chan result, requestList *guardedLogs, sctList *guardedLogs, submitter Submitter) {
```
Naming of `requestList` and `sctList` is a bit confusing -- they're sets not lists, and they're sets of logs not SCTs nor requests.  Maybe `submittedTo` and `receivedSCTFrom` ?

Mercurrent replied:
Renamed to requested and receievedSCTFrom as sending a request doesn't guarantee submit init.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// sleepTime calculates duration for consequent call. For first paralleStart calls duration is 0, while every next one gets additional dur interval.
+func sleepTime(idx int, parallelStart int, dur time.Duration) time.Duration {
```
`sleepTime` is a bit too generic as a function name

Mercurrent replied:
Renamed to postInterval.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+		}
+	}
+	err := completenessError(groups, &sctMap, gCtx)
+	return collectedSCTs, err
```
Function comment should mention that it's possible to have a result and a non-`nil` err.

Mercurrent replied:
Done.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	var collectedSCTs []*AssignedSCT
+	var m sync.Mutex
```
Actually, I don't think the mutex is needed -- only a single thread accesses `collectedSCTs`

Mercurrent replied:
Removed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+type guardedLogs struct {
+	logURLs map[string]bool
+	m       sync.RWMutex
```
Renamed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	results chan result, requestList *guardedLogs, sctList *guardedLogs, submitter Submitter) {
```
?

daviddrysdale replied:
Ah, typo: `results` could be a `chan<- result` -- this function only ever needs to write to the channel.

Mercurrent replied:
Changed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+		select {
+		case <-ctx.Done():
+			err := completenessError(groups, &sctMap, gCtx)
+			return collectedSCTs, err
```
We've already had this discussion. It's not intended to lose any SCTs collected. Even when error occurs.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+			}
+			// verify SCT
+			sctList.add(logURL)
+			results <- result{sct: sct, logURL: logURL}
```
Moved out SCT-collection.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+func parallelNums(groups map[string]*ctpolicy.LogGroupInfo) map[string]int {
```
Renamed to LogPolicyData.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+func completenessError(groups map[string]*ctpolicy.LogGroupInfo, sctMap *guardedLogs, gCtx map[string]groupCtx) error {
+	cmplt := groupsComplete(groups, sctMap, gCtx)
+	if !cmplt {
```
Done.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+type groupCtx struct {
+	ctx    context.Context
+	cancel context.CancelFunc
```
Got rid of that construction.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	for groupName, g := range groups {
+		if sctList.sumFor(g.LogURLs) >= g.MinInclusions {
+			if gCtx[groupName].ctx.Err() != nil {
+				gCtx[groupName].cancel()
```
Removed the whole construction.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+type groupCtx struct {
+	ctx    context.Context
```
Removed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	logURL string
+}
+
+func groupsComplete(groups map[string]*ctpolicy.LogGroupInfo, sctList *guardedLogs, gCtx map[string]groupCtx) bool {
```
Moved ctx-mess out of PR.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+	for _, g := range groups {
+		subCtx, cancel := context.WithCancel(ctx)
+		gCtx[g.Name] = groupCtx{ctx: subCtx, cancel: cancel}
+		defer gCtx[g.Name].cancel()
```
Had some vague idea. Now removing and left only one.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,215 @@
+
+	var collectedSCTs []*AssignedSCT
+	var m sync.Mutex
+	for i := 0; i < logsNumber; i++ {
```
It doesn't require. But it expects to process at most logsNumber of results.

Mercurrent replied:
Comment added.

Mercurrent replied:
* requested should belong to the requestor. Otherwise we'll have to introduce sessions/cert-trackers into Submitter.
* if receivedSCTFrom is local, then we have to (re)introduce sub-contexts for each of groups for proper management.

Mercurrent commented:
PTAL

Mercurrent commented on submission/races_test.go:
```diff
@@ -0,0 +1,152 @@
+type mockSubmitter struct {
+	// keeps track of number of requests per log-group. Logs splitted into groups based on logURL first letter.
+	firstLetterURLReqNumber map[byte]int
+	m                       sync.Mutex
```
Done.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+)
+
+type safeStringSet struct {
+	logURLs map[string]bool
```
Renamed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+	mu      sync.Mutex
+}
+
+// Returns whether add was successful: no entry before, entry added.
```
Rewrote comment.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+func (gList *safeStringSet) has(logURL string) bool {
+	gList.mu.Lock()
+	defer gList.mu.Unlock()
+	exist := gList.logURLs[logURL]
```
Done.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+	return nums
+}
+
+func totalLogs(groups ctpolicy.LogPolicyData) int {
```
While I agree reg. totalLogs (moved), I'm in doubt reg. the parallelCalls one. As it's more related to the submit logic, rather than to the policy itself.

Mercurrent commented on submission/races_test.go:
```diff
@@ -0,0 +1,152 @@
+		name        string
+		sbMock      Submitter
+		groups      ctpolicy.LogPolicyData
+		ctx         context.Context
```
For now. I have intention for some more tests.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+func completenessError(groups ctpolicy.LogPolicyData, receivedSCTFrom *safeStringSet) error {
+	if cmplt := groupsComplete(groups, receivedSCTFrom); !cmplt {
```
Fixed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+				return
+			}
+			// verify SCT
+			results <- result{sct: sct, logURL: logURL}
```
Swapped.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,208 @@
+
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	results chan result, requested *safeStringSet, receivedSCTFrom *safeStringSet, submitter Submitter) {
+	groupURLs := make([]string, len(group.LogURLs))
```
Done.

daviddrysdale commented:
Couple of process things:

 - Once a PR is out for review, it's better to avoid re-writing Git history (to avoid comments being attached to commits that disappear) -- could you stick to adding 'fixup' or 'review markup' commits at the end of the branch?
 - The GitHub UI shows a few unaddressed comments from earlier rounds -- are you planning to fix or reply to them?

Mercurrent commented:
- added comment for log-permutations
- added returning possible error alongside SCT
- added comment into cycle through results clarifying finalize-conditions.

daviddrysdale commented on submission/races_test.go:
```diff
@@ -0,0 +1,152 @@
+			}
+		})
+	}
+}
```
Could do with a bunch more test cases here, but probably worth holding off until we're settled on the mainline code.  

Things to possibly test:
 - `safeStringSet` behaviour
 - individual methods like `postInterval()`, `parallelNums()`, `groupsComplete()`
 - more variations of submission failures, timeouts, log counts etc (with the aim of hitting every possible exit condition from `groupRace` and from the loop in `GetSCTs`)

Mercurrent replied:
I thought the comment on the very beginning of this review reflects why there's not that many tests (yet).

daviddrysdale replied:
Ah yes, I forgot about that.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,196 @@
+			if receivedSCTFrom.sumFor(group.LogURLs) >= group.MinInclusions {
+				return
+			}
+			sct, err := submitter.SubmitToLog(ctx, logURL, chain)
```
Ok, removed wrappers.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,196 @@
+	for _, g := range groups {
+		go func(group *ctpolicy.LogGroupInfo) {
+			groupRace(subCtx, chain, group, parallelNums[group.Name], results, &requested, &receivedSCTFrom, submitter)
+		}(g)
```
reasonable question

Mercurrent commented on submission/races_test.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+type mockSubmitter struct {
+	// keeps track of number of requests per log-group. Logs splitted into groups based on logURL first letter.
```
Done.

Mercurrent commented:
Addressed re-naming/re-structure.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,187 @@
+				return
+			}
+			sct, err := submitter.SubmitToLog(ctx, logURL, chain)
+			// verify SCT
```
Done.

daviddrysdale commented:
As previously mentioned, it's easier for your reviewer if you stick to adding new commits to the branch rather than rebasing and force-pushing.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,227 @@
+		go groupRace(subCtx, chain, g, parallelNums[g.Name], &submissions, groupEvents, submitter)
+	}
+
+	groupComplete := make(map[string]bool)
```
Could do with a comment here, as this isn't just a set -- IIUC something like:
```golang
// A group name being present in the map indicates that group has
// completed processing, and the bool value indicates whether it
// completed successfully.
```

Mercurrent replied:
Done.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,227 @@
+		}
+	}
+	if len(failedGroups) > 0 {
+		return errors.New("Log-group(s) " + strings.Join(failedGroups, ", ") + "didn't receive enough SCTs.")
```
style: Use `fmt.Errorf` and start error messages with a lower-case letter:
```golang
return fmt.Errorf("log-group(s) %s didn't receive enough SCTs.", strings.Join(failedGroups, ", "))
```

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,227 @@
+	submissions := safeSubmissionSet{results: make(map[string]*submissionResult)}
+	// channel listening to group-completion (failure) events
+	groupEvents := make(chan groupState, len(groups))
+	subCtx, cancel := context.WithCancel(ctx)
```
Is `subCtx` needed here?  Its `cancel` only fires at function exit, which can only happen if all groups are finished already (so no need to cancel) or if the master context is done (in which case just using `ctx` would terminate things too).

Mercurrent replied:
Changed into subCtx per Log-request.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,227 @@
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
```
`counter` looks it could be replace with a `sync.WaitGroup`:
```golang
	var wg sync.WaitGroup
	for i, urlNum := range rand.Perm(len(groupURLs)) {
		wg.Add(a)
		go func(i int, logURL string) {
			defer wg.Done()
			...
		}(i, groupURLs[urlNum])
	}
	// Wait until all per-group goroutines are done.
	wg.Wait()
	...
```
However, this will need the `Sleep` to be interruptible by a cancelled (sub-)context -- see below.

Mercurrent replied:
Still need ability to check completion on each counter.

daviddrysdale replied:
This still looks like it should be a `sync.WaitGroup` to me.  Also, if you have an invariant that the inner goroutine must always do something on exit (whether it's `counter <- count{}` or `wg.Done()`), then use a `defer`.

Mercurrent replied:
The logic has been updated since.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,227 @@
+			if ctx.Err() != nil {
+				counter <- count{}
+				return
+			}
```
It would be better if the `Sleep` here were interruptible by a cancelled context -- something like:
```golang
	ticker := time.NewTicker(postInterval(i, parallelStart, PostBatchInterval))
	defer ticker.Stop()
	select {
	case <- ctx.Done():
		return
	case <- ticker.C:
	}
```

Mercurrent replied:
I don't need a regular ticker. Added timeAfter.

daviddrysdale replied:
Yeah, I forgot about that -- `time.After` makes more sense.

daviddrysdale commented on submission/races.go:
```diff
@@ -0,0 +1,227 @@
+			if state.sumSCTsFor(group.LogURLs) >= group.MinInclusions {
+				completionMu.Lock()
+				if !earlyCompletion {
+					completion <- groupState{Name: group.Name, Success: true}
```
If the sleep above is interruptible, I *think* this special case could be dropped, in exchange for using a sub-context.  This arm could just early-invoke the cancel so all the other goroutines would drop out and the `wg.Wait()` would finish, then the last line of the function writes the result to `completion`.

That in turn would allow a further simplification: don't pass in the `completion` channel at all, just have this function return a `groupState`.  Then the caller could do:
```golang
	for _, g := range groups {
		go func(g *ctpolicy.LogGroupInfo) {
			groupEvents <- groupRace(subCtx, chain, g, parallelNums[g.Name], &submissions, submitter)
		}(g)
	}
```
and the `groupEvents` channel never has to be passed around.

Mercurrent replied:
Got rid of passing groupEvents channel.

Mercurrent commented:
Adding Pierre as he volunteered.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package races
```
Does it have to? Renamed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	needs, ok := sub.groupNeeds[groupName]
+	if !ok {
```
Well. If the group is unknown, then it's up to heuristics whether it's complete. I'd treat it as non-blocking, thus as complete.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+		sub.groupNeeds[groupName]--
+	}
+
+	// Cancel any pending Log-requests for which there's no more awaiting Log-group.
```
Fixed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+	return sub.results[logURL] != nil
+}
+
+// Gets SCT-result. Writes it down if it is error or awaited-SCT. Re-calculates group-completion and cancels any running but not-awited-anymore Log-requests.
```
Updated, fixed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+}
+
+// returns whether the Log has already been requested.
+func (sub *safeSubmissionState) has(logURL string) bool {
```
Not in this PR. removed.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+	return &s
+}
+
+// request includes empty submissionResult in the set, returns whether the entry wasn't requested (added) before.
```
Updated the comment.

Mercurrent commented on submission/races.go:
```diff
@@ -0,0 +1,284 @@
+
+	results map[string]*submissionResult
+	cancels map[string]context.CancelFunc
+	mu      sync.Mutex
```
Moved.
